soccer/radio/NetworkRadio.cpp

```diff
@@ -0,0 +1,187 @@
+    : _socket(_context, udp::endpoint(udp::v4(), server_port)),
+      _server_port{server_port},
+      _robot_port{robot_port} {}
+
+NetworkRadio::~NetworkRadio() {}
+
+bool NetworkRadio::open() {
+    startReceive();
+    return true;
+}
+
+void NetworkRadio::startReceive() {
+    // Set a receive callback
+    _socket.async_receive_from(
+        boost::asio::buffer(_recv_buffer), _robot_endpoint,
+        [this] (const boost::system::error_code& error, std::size_t num_bytes) {
+            receivePacket(error, num_bytes);
+        });
+}
+
+bool NetworkRadio::isOpen() const {
+    return _socket.is_open();
+}
+
+void NetworkRadio::send(Packet::RadioTx& packet) {
+    // Get a list of all the IP addresses this packet needs to be sent to
+    for (int robot_idx = 0; robot_idx < packet.robots_size(); robot_idx++) {
+        const Packet::Control& control = packet.robots(robot_idx).control();
+        uint32_t robot_id = packet.robots(robot_idx).uid();
+
+        // Build the control packet for this robot.
+        std::array<uint8_t, rtp::HeaderSize + sizeof(rtp::RobotTxMessage)> forward_packet;
+
+        rtp::Header* header = reinterpret_cast<rtp::Header*>(&forward_packet[0]);
+        header->port = rtp::PortType::CONTROL;
+        header->address = rtp::BROADCAST_ADDRESS;
+        header->type = rtp::MessageType::CONTROL;
+
+        size_t offset = sizeof(rtp::Header);
+        rtp::RobotTxMessage* msg = reinterpret_cast<rtp::RobotTxMessage*>(
+                &forward_packet[offset]);
+
+        // Set fields of the control message.
+        msg->uid = robot_id;
+
+        msg->message.controlMessage.bodyX = static_cast<int16_t>(
+            control.xvelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.bodyY = static_cast<int16_t>(
+            control.yvelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.bodyW = static_cast<int16_t>(
+            control.avelocity() * rtp::ControlMessage::VELOCITY_SCALE_FACTOR);
+        msg->message.controlMessage.dribbler =
+            clamp(static_cast<uint16_t>(control.dvelocity()) * 2, 0, 255);
+
+        msg->message.controlMessage.shootMode = control.shootmode();
+        msg->message.controlMessage.kickStrength = control.kcstrength();
+        msg->message.controlMessage.triggerMode = control.triggermode();
+        msg->message.controlMessage.song = control.song();
+
+        // Fetch the IP address
+        auto range = _robot_ip_map.left.equal_range(robot_id);
+
+        // Loop over all addresses for this robot; we want to send multiple copies.
+        for (auto it = range.first; it != range.second; it++) {
+            // Send to the given IP address
+            ip::address robot_ip = it->second;
+            udp::endpoint _robot_endpoint(robot_ip, _robot_port);
+            _socket.async_send_to(boost::asio::buffer(forward_packet), _robot_endpoint,
```

Comment:
I'm worried about the `forward_packet` object not remaining valid until the handler is called. I don't think we have a guarantee that it is

Reply:
Good catch. Not sure how I missed that, this is entirely wrong.